packages/cli-platform-android/src/commands/buildAndroid/index.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  CLIError,
+  getDefaultUserTerminal,
+  isPackagerRunning,
+  logger,
+} from '@react-native-community/cli-tools';
+import {Config} from '@react-native-community/cli-types';
+import execa from 'execa';
+import {getAndroidProject} from '../../config/getAndroidProject';
+import {getTaskNames, toPascalCase} from '../runAndroid/runOnAllDevices';
+import adb from '../runAndroid/adb';
+import getAdbPath from '../runAndroid/getAdbPath';
+import {startServerInNewWindow} from '../runAndroid';
+
+export interface BuildFlags {
+  mode: 'debug' | 'release';
+  variant?: string;
+  activeArchOnly?: boolean;
+  packager: boolean;
+  port: number;
+  terminal: string;
+  tasks?: Array<string>;
+}
+
+export async function checkPackager(args: BuildFlags, config: Config) {
+  if (!args.packager) {
+    return Promise.resolve();
+  }
+  const result = await isPackagerRunning(args.port);
+  if (result === 'running') {
+    logger.info('JS server already running.');
+  } else if (result === 'unrecognized') {
+    logger.warn('JS server not recognized, continuing with build...');
+  } else {
+    // result == 'not_running'
+    logger.info('Starting JS server...');
+    try {
+      startServerInNewWindow(args.port, args.terminal, config.reactNativePath);
+    } catch (error) {
+      logger.warn(
+        `Failed to automatically start the packager server. Please run "react-native start" manually. Error details: ${error.message}`,
+      );
+    }
+  }
+  return Promise.resolve();
+}
+
+async function buildAndroid(
+  _argv: Array<string>,
+  config: Config,
+  args: BuildFlags,
+) {
+  const androidProject = getAndroidProject(config);
+  const variant = args.variant ?? (args.mode || 'debug');
+  const tasks = args.tasks || ['assemble' + toPascalCase(variant)];
+  const gradleArgs = getTaskNames(androidProject.appName, tasks);
+
+  const adbPath = getAdbPath();
+  const devices = adb.getDevices(adbPath);
+
+  if (args.activeArchOnly) {
```

Comment:
Let's defer running this code
```suggestion
  if (args.activeArchOnly) {
    const adbPath = getAdbPath();
    const devices = adb.getDevices(adbPath);
```